conv_rnn/train.py

```diff
@@ -0,0 +1,192 @@
+    n_epochs = kwargs.get("n_epochs", 30)
+    restore = kwargs.get("restore", False)
+    verbose = kwargs.get("verbose", True)
+    lr = kwargs.get("lr", 5E-2)
+    weight_decay = kwargs.get("weight_decay", 1E-3)
+    schedule_factor = kwargs.get("schedule_factor", 0.2)
+    gradient_clip = kwargs.get("gradient_clip", 5)
+    seed = kwargs.get("seed", 3)
+    patience = kwargs.get("patience", 20)
+
+    if not kwargs["no_cuda"]:
+        torch.cuda.set_device(1)
+    set_seed(seed)
+    data_loader = data.SSTDataLoader("data")
+    if restore:
+        conv_rnn = torch.load(kwargs["input_file"])
+    else:
+        id_dict, weights, unk_vocab_list = data_loader.load_embed_data()
+        word_model = model.SSTWordEmbeddingModel(id_dict, weights, unk_vocab_list)
+        if not kwargs["no_cuda"]:
+            word_model.cuda()
+        conv_rnn = model.ConvRNNModel(word_model, **kwargs)
+        if not kwargs["no_cuda"]:
+            conv_rnn.cuda()
+
+    criterion = nn.CrossEntropyLoss()
+    parameters = list(filter(lambda p: p.requires_grad, conv_rnn.parameters()))
+    optimizer = torch.optim.Adadelta(parameters, lr=lr, weight_decay=weight_decay)
+    train_set, dev_set, test_set = data_loader.load_sst_sets()
+    scheduler = ReduceLROnPlateau(optimizer, mode="max", factor=schedule_factor, patience=patience)
+
+    for epoch in range(conv_rnn.epoch, n_epochs):
+        conv_rnn.train()
+        optimizer.zero_grad()
+        np.random.shuffle(train_set)
+        print("Epoch number: {}".format(epoch), end="\r")
+        if verbose:
+            print()
+        i = 0
+        while i + mbatch_size < len(train_set):
+            if verbose and i % (mbatch_size * 10) == 0:
+                print("{} / {}".format(i, len(train_set)), end="\r")
+            mbatch = train_set[i:i + mbatch_size]
+            train_in, train_out = conv_rnn.convert_dataset(mbatch)
+
+            if not kwargs["no_cuda"]:
+                train_in.cuda()
+                train_out.cuda()
+
+            scores = conv_rnn(train_in)
+            loss = criterion(scores, train_out)
+            loss.backward()
+            torch.nn.utils.clip_grad_norm(parameters, gradient_clip)
+            optimizer.step()
+            i += mbatch_size
+            if i % (mbatch_size * 256) == 0:
+                conv_rnn.eval()
+                dev_in, dev_out = conv_rnn.convert_dataset(dev_set)
+                scores = conv_rnn(dev_in)
+                n_correct = (torch.max(scores, 1)[1].view(len(dev_set)).data == dev_out.data).sum()
+                accuracy = n_correct / len(dev_set)
+                #scheduler.step(accuracy)
+                if accuracy > conv_rnn.best_dev:
+                    conv_rnn.best_dev = accuracy
+                    torch.save(conv_rnn, kwargs["output_file"])
+                if verbose:
+                    print("Dev set accuracy: {}".format(accuracy))
+                conv_rnn.train()
+    return conv_rnn.best_dev
+
+def do_random_search():

```

Comment:
Where is this `do_random_search` and ` do_grid_search` called? Please add some driver code that trigger these functions.